dbms/src/Columns/ColumnString.cpp

```diff
@@ -379,35 +421,73 @@ void ColumnString::updateWeakHash32(WeakHash32 & hash, const TiDB::TiDBCollatorP
     if (hash.getData().size() != s)
         throw Exception(fmt::format("Size of WeakHash32 does not match size of column: column size is {}, hash size is {}", s, hash.getData().size()), ErrorCodes::LOGICAL_ERROR);
 
-    const UInt8 * pos = chars.data();
     UInt32 * hash_data = hash.getData().data();
-    Offset prev_offset = 0;
 
     if (collator != nullptr)
     {
-        for (const auto & offset : offsets)
+        if (collator->getCollatorId() == TiDB::ITiDBCollator::UTF8MB4_BIN)
         {
-            auto str_size = offset - prev_offset;
-            /// Skip last zero byte.
-            auto sort_key = collator->sortKey(reinterpret_cast<const char *>(pos), str_size - 1, sort_key_container);
-            *hash_data = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(sort_key.data), sort_key.size, *hash_data);
-
-            pos += str_size;
-            prev_offset = offset;
+            // Skip last zero byte.
+            LoopOneColumn(chars, offsets, offsets.size(), [&](const std::string_view & view, size_t) {
+                auto sort_key = BinCollatorSortKey<true>(view.data(), view.size());
+                *hash_data = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(sort_key.data), sort_key.size, *hash_data);
+                ++hash_data;
+            });
+        }
+        else
+        {
+            // Skip last zero byte.
+            LoopOneColumn(chars, offsets, offsets.size(), [&](const std::string_view & view, size_t) {
+                auto sort_key = collator->sortKey(view.data(), view.size(), sort_key_container);
+                *hash_data = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(sort_key.data), sort_key.size, *hash_data);
+                ++hash_data;
+            });
+        }
+    }
+    else
+    {
+        // Skip last zero byte.
+        LoopOneColumn(chars, offsets, offsets.size(), [&](const std::string_view & view, size_t) {
```

Comment:
Both `updateHashWithValues` and `updateHashWithValue` don't remove the last zero byte if `collator == nullptr`, I think we should also keep the last zero byte here, otherwise, it may make the others confused.

Reply:
No change here. https://github.com/pingcap/tiflash/blob/99d9eb0646c077cbb186e8b8b31ae96f705bff5f/dbms/src/Columns/ColumnString.cpp#L484-L491.

The web of github show mistake.